src/Nethermind/Nethermind.Synchronization/ParallelSync/MultiSyncModeSelector.cs

```diff
@@ -332,6 +336,24 @@ private bool ShouldBeInWaitingForBlockMode(Snapshot best)
             return result;
         }
 
+        private bool ShouldBeInBeaconHeaders()
+        {
+            return _beaconSyncStrategy.ShouldBeInBeaconHeaders() &&
+                   !ShouldBeInUpdatingPivot();
+        }
+
+        private bool ShouldBeInUpdatingPivot()
+        {
+            bool updateRequestedAndNotFinished = _syncConfig.MaxAttemptsToUpdatePivot > 0;
+            bool isPostMerge = _beaconSyncStrategy.GetFinalizedHash() != null;
+            bool stateSyncNotFinished = _syncProgressResolver.FindBestFullState() == 0;
+
+            return updateRequestedAndNotFinished &&
+                   FastSyncEnabled &&
+                   isPostMerge &&
+                   stateSyncNotFinished;
```

Comment:
I think if we even start state sync, there is no point of updating Pivot as we already FastSynced to Head?

Reply:
please add logger.Trace like for other methods

Reply:
We need it to have correct pivot when downloading headers/bodies/receipts to avoid gaps. After restart pivot is set back to value from config. We are going to `PivotUpdator`, it is checking db for updated values, founding it, setting pivot back to updated value and that's all. We just have logic of reading already updated pivot from db in class `PivotUpdator`.
Actually I think it can be problematic if restart would be during old bodies. Then we will not read it from db and sync of old receipts would start from old pivot. I think it is a bug and I need to change `stateSyncNotFinished` to `oldReceiptsNotFinished`, as we stop carrying about pivot only when we are fully synced (not only state). Or maybe even change it to always go to `PivotUpdator` and read updated pivot from db? Need to think about it a bit more

Reply:
and logger added

Reply:
> I think if we even start state sync, there is no point of updating Pivot as we already FastSynced to Head?

Right. We are trying to read already updated pivot from DB during initialization (in `InitializeNetwork` step). If it is successful, `updateRequestedAndNotFinished = _syncConfig.MaxAttemptsToUpdatePivot > 0` check will return `false` and mode would not change to `UpdatingPivot`. If not, we want to update pivot always if we are post-merge and fast sync in enabled. No need to check state here.

Reply:
Actually we need this state check - it is for already synced nodes which will be updated to version with `PivotUpdator`. We don't want to update pivot in that case.
Looks like it was implemented right way from the very beginning, I just forgot why was that during long period from writing it 😄

Reply:
So to sum up - right about

> I think if we even start state sync, there is no point of updating Pivot as we already FastSynced to Head?

but this check is needed for backward compatibility with nodes synced before updating to Nethermind version with `PivotUpdator`.

In general, pivot can be updated only once in node lifetime. Then is saved to DB and loaded during initialization.